\documentclass{bioinfo}
\copyrightyear{2015} \pubyear{2015}

\access{Advance Access Publication Date: Day Month Year}
\appnotes{Application Note}

\usepackage{url}

\begin{document}
\firstpage{1}

\subtitle{Application Note}

\title[Indel Historian: phylogenetic ancestral reconstruction]{Indel Historian: fast and accurate ancestral sequence reconstruction using weighted finite-state transducers}
\author[Ian Holmes]{Ian Holmes$^{\text{\sfb 1}*}$}
\address{$^{\text{\sf 1}}$Department of Bioengineering, University of California, Berkeley, 94703, USA.}

\corresp{$^\ast$To whom correspondence should be addressed.}

\history{Received on XXXXX } %; revised on XXXXX; accepted on XXXXX}

\editor{Associate Editor: XXXXXXX}

\abstract{
{\bf Motivation.}
{\bf Results.}
{\bf Availability and Implementation.}
Indel Historian is available at \url{https://github.com/ihh/indelhistorian} under the Creative Commons Attribution 3.0 US license.
It is written in C++11.
{\bf Contact.}
Ian Holmes {\tt ihholmes+indelhistorian@gmail.com}.
{\bf Supplementary Information.}
None.
}

\maketitle

\section{Introduction}

Synthesis of reconstructed ancestral proteins offers a direct way to test hypotheses of evolutionary adaptation \citep{UgaldeEtAl2004,OrtlundEtAl2007,GaucherEtAl2008}
or to generate diversity in combinatorial libraries for bioengineering by directed evolution \citep{AlcolombriEtAl2011,SantiagoOrtizEtAl2015}.
The computational workflow for such reconstructions typically involves building a multiple sequence alignment and then using statistical phylogenetics to reconstruct the substitution history
\citep{Liberles2007}.
However, it remains an open question whether the multiple alignment tools that perform best on the most commonly-used benchmarks
(which generally use protein structure as a gold standard)
are also the best alignment tools for performing ancestral sequence reconstruction.

Multiple sequence alignments are used for several purposes in bioinformatics,
and these different applications have led to different criteria for evaluating
the accuracy of the tools.
Homology modeling in protein structure prediction
leads to accuracy metrics that quantify the number of correctly aligned residues
compared to a benchmark dataset of structurally-informed reference alignments \citep{ThompsonEtAl2005}.
Techniques that have proven successful in this context include using Bayesian decision theory to maximize the expected
score under the metrics \citep{NotredameEtAl2000,DoEtAl2005,SchwartzPachter2007,BradleyEtAl2009}
and fine-tuning the alignment scoring function to reflect known details of
protein sequences under selection for folding, such as the reduced likelihood of indels in
hydrophobic regions \citep{KatohEtAl2005,Edgar2004b,LarkinEtAl2007}.
These have been accompanied by performance optimizations of the various steps involved in multiple alignment,
the rate-limiting step generally being all-versus-all sequence comparison \citep{BradleyEtAl2009,Edgar2004b}.

The study of sequence evolution and molecular phylogenetics gives rise to
a different set of criteria for measuring alignment accuracy.
Several independent studies have showed that, for the purposes of estimating
molecular evolutionary parameters---such as the rates of insertion-deletion events \citep{Westesson2012-zg},
the ratio between synonymous and nonsynonymous substitution rates \citep{Redelings2014},
or other evolutionary analyses \citep{LoytynojaGoldman2008}---it is best to use an explicit probabilistic model of the sequence evolution process
and to treat sequence alignment rigorously as a missing data estimation problem
under this model.
One conceptual explanation for this phenomenon is that, for the purposes of predicting a protein structure,
functionally equivalent sequences in a rapidly-evolving region (such as a loop)
may be considered alignable even if they have been entirely replaced by a series of indel mutations
(e.g. the sequences $X$ and $Y$ in the mutation trajectory $AXB \to AXYB \to AYB$),
but for the purposes of evolutionary analysis, one generally wants to count every indel event that has occurred.

The ProtPal program \citep{Westesson2012-zg} is one such approach which uses the theory of
weighted finite-state transducers,
automata which can be multiplied together like substitution matrices \cite{BouchardCote2013}.
This approach can be represented as a strict generalization of Felsenstein's celebrated
algorithm for computing likelihoods on phylogenetic trees \citep{Felsenstein81}.
In fact, the use of transducers makes Felsenstein's pruning algorithm formally
equivalent to a special case of Sankoff's phylogenetic multiple alignment and RNA folding algorithm \citep{Sankoff85};
the RNA folding component is included if one generalizes from finite-state transducers to pushdown automata \citep{BradleyHolmes2009}.

Our earlier benchmark of alignment methods on simulated sequence data
showed that most alignment algorithms introduce systematic biases into the
estimation of indel rates, and that these biases are typically substantially worse
at higher indel rates \citep{Westesson2012-zg}.
ProtPal was found to be the most accurate algorithm for indel rate estimation,
followed by PRANK and then MUSCLE.
However, the implementation of ProtPal published with that benchmark was too slow
for practical use.
Furthermore, no benchmark of ProtPal on structural alignment benchmarks was available,
and---although we have noted that structure prediction is a different task from evolutionary analysis---it would be
useful to know just how different it is, according to the metrics.

Here, we present a clean implementation of the algorithm underlying ProtPal
in a new tool, Indel Historian.
The tool is targeted to evolutionary applications such
as ancestral sequence reconstruction and molecular phylogenetics.
We also report an assessment of the structural alignment accuracy of IndelHistorian on the benchmark datasets BAliBase 3.0 \citep{ThompsonEtAl2005}
and PREFAB 4.0 \citep{Edgar2010}.

\begin{methods}
\section{Methods}

\citep{Westesson2012-zg}

\end{methods}



Comparison to MCMC tools
BaliPhy \citep{RedelingsSuchard2005,RedelingsSuchard2007,Redelings2014},
StatAlign \citep{NovakEtAl2008,HermanEtAl2014},
HandAlign \citep{WestessonBarquistHolmes2012}


\section{Results}



\section{Discussion}



The best-performing method MUSCLE introduces heuristics into its scoring scheme
that improve accuracy.

It is harder to incorporate these sorts of modifications into a rigorous phylogenetic model
primarily because they violate the assumption of independence
between indel and substitution processes
that most such models rely on

Decision theory may also be useful to produce consensus alignments summarizing an MCMC run \citep{HermanEtAl2015}



\section{Availability}

Indel Historian is available at \url{https://github.com/ihh/indelhistorian} under the Creative Commons Attribution 3.0 US license. It is written in C++11, and compiles on a POSIX system with Clang (v.6.1.0). It requires the GNU Scientific Software library (libgsl), the ZLib compression library (libz), and the Boost C++ library.

\section*{Acknowledgements}

Indel Historian includes code from Ivan Vashchaev (gason), Heng Li (kseq), Rafael Baptista (stacktrace), David Robert Nadeau (memsize), Roger Pate (string escaping), Konrad Rudolph (index sort), and the StackOverflow community.

\section*{Funding}

This work has been supported by NHGRI grant R01-HG004483.

\bibliographystyle{natbib}
%\bibliographystyle{bioinformatics}
%\bibliographystyle{achemnat}
%\bibliographystyle{plainnat}
%\bibliographystyle{abbrv}
%
%\bibliographystyle{plain}
%
%\bibliography{Document}


\bibliography{references}
\end{document}
