\documentclass{bioinfo}
\copyrightyear{2015} \pubyear{2015}

\access{Advance Access Publication Date: Day Month Year}
\appnotes{Application Note}

\usepackage{url}

\begin{document}
\firstpage{1}

\subtitle{Application Note}

\title[Indel Historian: phylogenetic ancestral reconstruction]{Indel Historian: fast and accurate ancestral sequence reconstruction}
\author[Ian Holmes]{Ian Holmes$^{\text{\sfb 1}*}$}
\address{$^{\text{\sf 1}}$Department of Bioengineering, University of California, Berkeley, 94703, USA.}

\corresp{$^\ast$To whom correspondence should be addressed.}

\history{Received on XXXXX } %; revised on XXXXX; accepted on XXXXX}

\editor{Associate Editor: XXXXXXX}

\abstract{
{\bf Motivation.}
Ancestral reconstruction of indel histories is a specialized application of multiple sequence alignment.
The alignment tool ProtPal, which uses weighted finite-state transducers to model indel evolution,
out-performed other ancestral-reconstruction alignment tools in a previous simulation benchmark,
but has been too slow for practical use.
{\bf Results.}
Indel Historian combines an efficient reimplementation of the ProtPal algorithm with performance-improving heuristics from other alignment tools.
On structural alignment benchmarks, it out-performs PRANK and is comparable to CLUSTALW.
{\bf Availability and Implementation.}
Indel Historian is available at \url{https://github.com/ihh/indelhistorian} under the Creative Commons Attribution 3.0 US license.
{\bf Contact.}
Ian Holmes {\tt ihholmes+indelhistorian@gmail.com}.
{\bf Supplementary Information.}
None.
}

\maketitle

\section{Introduction}

Synthesis of reconstructed ancestral proteins offers a direct way to test hypotheses of evolutionary adaptation \citep{UgaldeEtAl2004,OrtlundEtAl2007,GaucherEtAl2008}
or to generate diversity in combinatorial libraries for directed evolution \citep{AlcolombriEtAl2011,SantiagoOrtizEtAl2015}.
A typical computational workflow involves building a multiple sequence alignment and then using statistical phylogenetics to reconstruct the substitution history
\citep{Liberles2007}.
However, it remains unclear whether the multiple alignment tools that perform best on common benchmarks
(which generally use protein structure as a gold standard)
are also the best tools for ancestral reconstruction.

Multiple alignments are used for several purposes in bioinformatics,
only one of which is homology-directed structure prediction,
yet structurally-informed alignments have tended to dominate alignment benchmarks.
Alignment accuracy scores (such as SPS, TCS, AMA, and the Cline shift score) quantify the number of correctly aligned residues
compared to a benchmark dataset of structurally-informed reference alignments \citep{ThompsonEtAl2005}.
Techniques that score highly in these tests include using Bayesian decision theory to maximize the expected
alignment accuracy \citep{NotredameEtAl2000,DoEtAl2005,SchwartzPachter2007,BradleyEtAl2009}
and fine-tuning the alignment scoring function for known quirks of
protein evolution, such as the reduced likelihood of indels in
hydrophobic regions \citep{KatohEtAl2005,Edgar2004b,LarkinEtAl2007}.
These have been accompanied by performance optimizations of 
rate-limiting steps such as all-versus-all pairwise sequence comparison \citep{BradleyEtAl2009,Edgar2004b}.

There is evidence that for evolutionary uses of alignments, such as reconstructing trees or ancestral sequences,
different tools (or at least different tool-assessment metrics) might be required.
Several independent studies have shown that, for the purposes of estimating
molecular evolutionary parameters---such as indel rates \citep{Westesson2012-zg},
dN/dS ratios \citep{Redelings2014},
or trees \citep{LoytynojaGoldman2008}---it is best to use an explicit probabilistic model of the sequence evolution process
and to treat sequence alignment rigorously as a statistical ``missing data'' problem.
One possible explanation for this is that, for protein structure prediction,
detailed reconstruction of indel histories in fast-evolving regions (such as loops) is unnecessary,
but evolutionary analyses can make more use of this information.

The ProtPal program \citep{Westesson2012-zg} is one such approach, modeling indel evolution using automata.
weighted finite-state transducers,
automata which can be multiplied together like substitution matrices \citep{BouchardCote2013},
generalizing Felsenstein's celebrated
algorithm for computing phylogenetic likelihoods from individual sites to whole sequences \citep{Felsenstein81}.
%In fact, the use of transducers makes Felsenstein's pruning algorithm formally
%equivalent to a special case of Sankoff's phylogenetic multiple alignment and RNA folding algorithm \citep{Sankoff85};
%the RNA folding component is included if one generalizes from finite-state transducers to pushdown automata \citep{BradleyHolmes2009}.
In an earlier benchmark of alignment methods on simulated sequence data,
ProtPal was found to be the most accurate algorithm for indel rate estimation,
followed by PRANK and then MUSCLE.
Most alignment algorithms introduce systematic biases into the
estimation of indel rates, and these biases are typically substantially worse
at higher indel rates \citep{Westesson2012-zg}.
These biases are minimized with ProtPal.
However, the implementation of ProtPal published with that benchmark was too slow
for practical use.
Furthermore, no benchmark of ProtPal on structural alignment benchmarks has been available,
and---although we have noted that structure prediction is a different task from evolutionary analysis---it would be
useful to know just how different it is, according to the metrics.

Here, we present a clean implementation of the algorithm underlying ProtPal
in a new tool, Indel Historian.
The tool is targeted to evolutionary applications such
as ancestral sequence reconstruction and molecular phylogenetics.
We also report an assessment of the alignment accuracy of IndelHistorian on structurally-informed benchmarks.

\begin{methods}
\section{Methods}

Indel Historian combines established algorithms from several sources.
Like ProtPal, Indel Historian progressively climbs a tree from tips to root,
using a finite-state transducer model of sequence evolution (similar to a Pair HMM)
to build an ancestral sequence profile that includes suboptimal alignments \citep{Westesson2012-zg}.
To find the guide tree, Indel Historian performs neighbor-joining \citep{SaitouNei87}
on a guide alignment constructed via a sparse random subset of the all-vs-all pairwise alignment graph \citep{BradleyEtAl2009}.
The progressive alignment step is followed by iterative refinement \citep{HolmesBruno2001,Edgar2004b} and optional MCMC \citep{WestessonBarquistHolmes2012}.
Indel Historian also includes implementations of the phylogenetic EM algorithm for continuous-time Markov chains \citep{HolmesRubin2002},
allowing substitution and indel rates to be estimated directly from sequence data.

\end{methods}

\section{Results}

\begin{table}
  \begin{tabular}{r|rr|rr}
    & \multicolumn{2}{c|}{BAliBase 3.0} & \multicolumn{2}{c}{PREFAB 4} \\
    & Mean SPS & Mean TCS & Mean SPS & Mean TCS \\
    \hline
IndelHistorian & 0.82 & 0.50 & 0.60 & 0.60 \\
CLUSTALW & 0.79 & 0.45 & 0.62 & 0.62 \\
PRANK & 0.73 & 0.35 & 0.51 & 0.51 \\
MUSCLE & 0.89 & 0.64 & 0.73 & 0.73 \\
  \end{tabular}
  \caption{
    Comparison of Indel Historian to other leading alignment programs using standard benchmarks
    and the SPS/TCS scores \citep{ThompsonEtAl2005}.
    Data for PRANK, CLUSTALW and MUSCLE are from {\tt http://drive5.com/bench/} \citep{Edgar2010}.
  }
\end{table}

Supplementing the previously published data on the superior fidelity of the ProtPal algorithm
at reconstructing indel histories \citep{Westesson2012-zg},
Table~1 summarizes an evaluation of the ProtPal algorithm, as implemented in Indel Historian, on structural benchmarks.
The comparison uses the BAliBase and PREFAB
benchmarks, compiled partially or wholly using 3D structure information.
In general, Indel Historian performs better than PRANK,
a program which also aims for a phylogenetically meaningful alignment
including ancestral sequences \citep{LoytynojaGoldman2008},
and the second-most accurate after ProtPal at estimating indel rates \citep{Westesson2012-zg}.
Compared to leading protein aligners that do not attempt ancestral reconstruction, Indel Historian performs better than CLUSTALW but worse than MUSCLE.

% The best-performing method MUSCLE introduces heuristics into its scoring scheme
% that improve accuracy.
% 
% It is harder to incorporate these sorts of modifications into a rigorous phylogenetic model
% primarily because they violate the assumption of independence
% between indel and substitution processes
% that most such models rely on
% 
% Comparison to MCMC tools
% BaliPhy \citep{RedelingsSuchard2005,RedelingsSuchard2007,Redelings2014},
% StatAlign \citep{NovakEtAl2008,HermanEtAl2014},
% HandAlign \citep{WestessonBarquistHolmes2012}
% 
% Decision theory may also be useful to produce consensus alignments summarizing an MCMC run \citep{HermanEtAl2015}



\section{Availability}

Indel Historian is available at \url{https://github.com/ihh/indelhistorian} under the Creative Commons Attribution 3.0 US license. It is written in C++11, and compiles on a POSIX system with Clang (v.6.1.0). It requires the GNU Scientific Software library (libgsl), the ZLib compression library (libz), and the Boost C++ library.

\section*{Acknowledgements}

Indel Historian includes code from Ivan Vashchaev (gason), Heng Li (kseq), % Rafael Baptista (stacktrace), David Robert Nadeau (memsize), Roger Pate (string escaping), Konrad Rudolph (index sort),
and the StackOverflow community.

\section*{Funding}

This work has been supported by NHGRI grant R01-HG004483.

\bibliographystyle{natbib}
%\bibliographystyle{bioinformatics}
%\bibliographystyle{achemnat}
%\bibliographystyle{plainnat}
%\bibliographystyle{abbrv}
%
%\bibliographystyle{plain}
%
%\bibliography{Document}


\bibliography{references}
\end{document}
